api/slideshow/api.go

```diff
@@ -0,0 +1,148 @@
+package slideshow
+
+import (
+	"errors"
+	"log"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pisign/pisign-backend/types"
+	"github.com/pisign/pisign-backend/utils"
+)
+
+type API struct {
+	types.BaseAPI
+	FilePathsForTags []string
+
+	Config types.SlideShowConfig
+	// Add other fields as necessary, (keep lowercase to avoid being stored in json)
+}
+
+// NewAPI creates a new API
+func NewAPI(sockets map[types.Socket]bool, pool types.Pool, id uuid.UUID) *API {
+	a := new(API)
+	a.BaseAPI.Init("slideshow", sockets, pool, id)
+
+	return a
+}
+
+// Configure based on client message
+func (a *API) Configure(message types.ClientMessage) error {
+	// Make sure the client widget is immediately sent new data with new config options
+	defer func() {
+		if a.Pool != nil && a.Sockets != nil {
+			a.Pool.Save()
+			a.Send(a.Data())
+		}
+	}()
+
+	if err := a.BaseAPI.Configure(message); err != nil {
+		return err
+	}
+
+	switch message.Action {
+	case types.ConfigureAPI, types.Initialize:
+		log.Printf("Configuring Slideshow:\n %s\n", string(message.Config))
+		if err := utils.ParseJSON(message.Config, &a.Config); err != nil {
+			log.Println("Could not properly configure Slideshow")
+			return errors.New("could not properly configure Slideshow")
+		}
+
+		// Add custom checks for config fields here (see the `time` api as an example)
+
+		log.Println("Slideshow configuration successful:", a)
+	}
+	log.Println("Slideshow config set to ", a.Config)
+
+	return nil
+}
+
+// Data performs a bulk of the computational logic
+func (a *API) Data() (interface{}, error) {
+	// Get the filepaths to include for a set of tags
+	// Also send all the possible tags so the client can choose later which tags to use
+	// when configuring the api
+	imgDB := a.Pool.GetImageDB().(*types.ImageDB)
+	images := make([]string, imgDB.NumImages-1)
+	uniqueTagsSet := make(map[string]bool)
+
+	for _, item := range imgDB.Images {
+		// Add to the tag to the tags set
+		added := false
+		for _, tag := range item.Tags {
+			uniqueTagsSet[tag] = true
+			if utils.StringInSlice(tag, a.Config.IncludedTags) {
+				if !added {
+					if len(item.FilePath[1:]) > 0 {
+						images = append(images, item.FilePath[1:])
+						added = true
+					}
+				}
+			}
+
+			// Include all? TODO add the custom ALL tag or something to also include it
+			if len(a.Config.IncludedTags) == 0 {
+				if len(item.FilePath[1:]) != 0 {
+					images = append(images, item.FilePath[1:])
+				}
+			}
+		}
+	}
+
+	uniqueTags := make([]string, len(uniqueTagsSet)-1)
+
+	// Get the unique tags from the set
+	for tag := range uniqueTagsSet {
+		uniqueTags = append(uniqueTags, tag)
+	}
+
+	// TODO there must be some issue with making slices that causes
+	// us to have empty lists... fix that so we don't have to rely on
+	// these functions to clean that up!
+	uniqueTags = utils.DeleteEmpty(uniqueTags)
+	imagesNotEmpty := utils.DeleteEmpty(images)
+
+	imgDB.UniqueTags = uniqueTags
```

Comment:
I feel like we shouldn't be editing any database data inside this function. Is there a specific reason this line needs to be here?

Reply:
Hmm, good thinking, probably only needs to be in the handler function for when they upload photos